### PR TITLE
Fix stuck escorts after traveling through a wormhole

### PR DESCRIPTION
**Pull request by yjhn**
**Bugfix:** Fixes these issues:
#2937
#4362 (tested with provided save file)
#3608 (tested with provided save file)
#4582 (tested with provided save file)
#5011 (tested with provided save file)
#5541 (tested)
the issue @Terin mentioned (tested)

## Fix Details
This PR fixes multiple issues with ship pathfinding:
1. ships not immediately following parent if they cannot jump or land on wormhole that the parent is targeting
       consider alternative travel plans (e.g. if parent jumps, but the escort cannot jump, or if the parent lands on wormhole but the escort cannot)
2. ships not following if the parent jumps while landing on wormhole
       reset jump orders while landing
3. ships getting stuck if they go through the wormhole off screen
       reset target system when going through wormhole
4. ships getting stuck with impossible destination if hyperspace link or system disappears
       reset destination system if it is unreachable

Also thanks to @oo13 for their excellent bug report #5541.

### CODE REVIEW

File: source/AI.cpp [hunk 1]
```diff
@@ -1433,7 +1433,7 @@ void AI::MoveIndependent(Ship &ship, Command &command) const
 		int jumps = ship.JumpsRemaining(false);
 		// Each destination system has an average priority of 10.
 		// If you only have one jump left, landing should be high priority.
-		int planetWeight = jumps ? (1 + 40 / jumps) : 1;
+		int planetWeight = jumps ? 41 : 1;
```
Comment (Amazinite) [inline source/AI.cpp:1436]: This is no longer doing what the comment suggests. Before, the less jumps a ship had remaining, the more likely it was to land. Now it just always has a high chance of landing.

File: source/AI.cpp [hunk 2]
```diff
@@ -1433,7 +1433,7 @@ void AI::MoveIndependent(Ship &ship, Command &command) const
 		int jumps = ship.JumpsRemaining(false);
```
Comment (Amazinite) [inline source/AI.cpp:1433]: We're now returning a bool to an int variable that we only ever use as a bool. (Lines 1436 and 1442 are just checking that jumps is non-zero.)

File: source/Ship.h [hunk 3]
```diff
@@ -280,7 +280,7 @@ class Ship : public Body, public std::enable_shared_from_this<Ship> {
 	// Get the number of jumps this ship can make before running out of fuel.
 	// This depends on how much fuel it has and what sort of hyperdrive it uses.
 	// If followParent is false, this ship will not follow the parent.
-	int JumpsRemaining(bool followParent = true) const;
+	bool JumpsRemaining(bool followParent = true) const;
```
Comment (Amazinite) [inline source/Ship.h:283]: The comment for this method is no longer accurate. We're no longer returning the number of jumps remaining, but whether the ship has any remaining jumps. Perhaps rename the method to HasJumpsRemaining.

File: source/AI.cpp (hunk 1 above)
Comment (tatami4) [reply source/AI.cpp:1436]: No, it isn't necessary, but it didn't return an accurate number of jumps before. It calculated the number of jumps left from only the fuel required for next jump.

**Comment by Terin** [comment]
[Stuck Remnant.txt](https://github.com/endless-sky/endless-sky/files/5811601/Stuck.Remnant.txt)

This is a save where mission escorts are stuck, not player ones, and this patch does not appear to fix it. Additionally, you can no longer ask for assistance for fuel, and you can start a jump without any fuel, and you will just sit and wait until you ramscoop enough.

**Comment by tatami4** [comment]
Also, this is unrelated, but interesting:
Ship 'Watchful Gravetender' of model 'Archon' in system 'Sayaiban' cannot reach its destination system 'Turais'
Ship 'Kasiliri's Folly' of model 'Archon' in system 'Fasitopfar' cannot reach its destination system 'Turais'
Ship 'Sword of Eden' of model 'Archon' in system 'Kasikfar' cannot reach its destination system 'Turais'
Ship 'Snows of Far Door Nai' of model 'Archon' in system 'Peresedersi' cannot reach its destination system 'Turais'
Ship 'R.N.S. Sangamon' of model 'Gunboat' in system 'Merak' cannot reach its destination system 'Merak'
Ship 'Sleeping Dragon' of model 'Archon' in system 'Nenia' cannot reach its destination system 'Turais'

**Comment by tatami4** [comment]
Escorts should no longer get stuck. Stuck escorts will still remain stuck (I will hopefully fix it soon).

**Comment by tatami4** [comment]
The issue @Terin mentioned is not fixed currently.

**Comment by tatami4** [comment]
All issues are now fixed. There is one caveat though: if the ships had an impossible destination (due to this bug), they will be stuck until their parent (usually the flagship) visits their system.

**Comment by Zitchas** [comment]
We just got another report of the bug this is supposed to solve turning up. Might we get this into the 0.9.15 milestone?

File: source/System.cpp [hunk 4]
```diff
@@ -829,7 +829,7 @@ void System::UpdateNeighbors(const Set<System> &systems, double distance)
 		// Skip systems that have no name.
 		if(it.first.empty() || it.second.Name().empty())
 			continue;
-
+
```
Comment (Hurleveur) [inline source/System.cpp:832]:
```suggestion

```

File: source/StarField.cpp [hunk 5]
```diff
@@ -327,7 +327,7 @@ void StarField::MakeStars(int stars, int width)
 	}
 	// Adjust the tile indices so that tileIndex[i] is the start of tile i.
 	tileIndex.insert(tileIndex.begin(), 0);
-
+
```
Comment (Hurleveur) [inline source/StarField.cpp:330]:
```suggestion

```

File: source/StarField.cpp [hunk 6]
```diff
@@ -200,7 +200,7 @@ void StarField::SetUpGraphics()
 		"  vec2 elongated = vec2(coord.x * size, coord.y * (size + elongation));\n"
 		"  gl_Position = vec4((rotate * elongated + translate + offset) * scale, 0, 1);\n"
 		"}\n";
-
+
```
Comment (Hurleveur) [inline source/StarField.cpp:203]:
```suggestion

```

File: source/StarField.cpp [hunk 7]
```diff
@@ -141,7 +141,7 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 				int count = 6 * tileIndex[index + 1] - first;
 				glDrawArrays(GL_TRIANGLES, first, count);
 			}
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:144]:
```suggestion

```

File: source/StarField.cpp [hunk 8]
```diff
@@ -125,7 +125,7 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 		// Round down to the start of the nearest tile.
 		minX &= ~(TILE_SIZE - 1l);
 		minY &= ~(TILE_SIZE - 1l);
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:128]:
```suggestion

```

File: source/StarField.cpp [hunk 9]
```diff
@@ -95,25 +95,25 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
...
 		GLfloat rotate[4] = {
 			static_cast<float>(unit.Y()), static_cast<float>(-unit.X()),
 			static_cast<float>(unit.X()), static_cast<float>(unit.Y())};
 		glUniformMatrix2fv(rotateI, 1, false, rotate);
-	
+		
 		glUniform1f(elongationI, length * zoom);
 		glUniform1f(brightnessI, min(1., pow(zoom, .5)));
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:116]:
```suggestion

```

File: source/StarField.cpp [hunk 10]
```diff
@@ -95,25 +95,25 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
...
 		GLfloat scale[2] = {baseZoom / Screen::Width(), -baseZoom / Screen::Height()};
 		glUniform2fv(scaleI, 1, scale);
-	
+		
 		GLfloat rotate[4] = {
 			static_cast<float>(unit.Y()), static_cast<float>(-unit.X()),
 			static_cast<float>(unit.X()), static_cast<float>(unit.Y())};
 		glUniformMatrix2fv(rotateI, 1, false, rotate);
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:113]:
```suggestion

```

File: source/StarField.cpp [hunk 11]
```diff
@@ -95,25 +95,25 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
...
 		// Don't zoom the stars at the same rate as the field; otherwise, at the
 		// farthest out zoom they are too small to draw well.
 		unit /= pow(zoom, .75);
-	
+		
 		float baseZoom = static_cast<float>(2. * zoom);
 		GLfloat scale[2] = {baseZoom / Screen::Width(), -baseZoom / Screen::Height()};
 		glUniform2fv(scaleI, 1, scale);
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:108]:
```suggestion

```

File: source/StarField.cpp [hunk 12]
```diff
@@ -95,25 +95,25 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
...
 		glBindVertexArray(vao);
-	
+		
 		float length = vel.Length();
 		Point unit = length ? vel.Unit() : Point(1., 0.);
 		// Don't zoom the stars at the same rate as the field; otherwise, at the
 		// farthest out zoom they are too small to draw well.
 		unit /= pow(zoom, .75);
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:104]:
```suggestion

```

File: source/StarField.cpp [hunk 13]
```diff
@@ -95,25 +95,25 @@ void StarField::Draw(const Point &pos, const Point &vel, double zoom) const
 	{
 		glUseProgram(shader.Object());
 		glBindVertexArray(vao);
-	
+
```
Comment (Hurleveur) [inline source/StarField.cpp:98]:
```suggestion

```

File: source/Ship.cpp [hunk 14]
```diff
@@ -3219,7 +3238,7 @@ void Ship::Jettison(const Outfit *outfit, int count)
 {
 	if(count < 0)
 		return;
-
+
```
Comment (Hurleveur) [inline source/Ship.cpp:3241]:
```suggestion

```

File: source/Fleet.cpp [hunk 15]
```diff
@@ -324,7 +325,7 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
 		
 		// Choose a random planet or star system to come from.
 		size_t choice = Random::Int(options);
-	
+
```
Comment (Hurleveur) [inline source/Fleet.cpp:328]:
```suggestion

```

File: source/Fleet.cpp [hunk 16]
```diff
@@ -299,15 +300,15 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
...
+		
 		// Find all the inhabited planets this fleet could take off from.
 		vector<const Planet *> planetVector;
 		if(!personality.IsSurveillance())
 			for(const StellarObject &object : system.Objects())
 				if(object.GetPlanet() && object.GetPlanet()->HasSpaceport()
 						&& !object.GetPlanet()->GetGovernment()->IsEnemy(government))
 					planetVector.push_back(object.GetPlanet());
-	
+
```
Comment (Hurleveur) [inline source/Fleet.cpp:311]:
```suggestion

```

File: source/Fleet.cpp [hunk 17]
```diff
@@ -299,15 +300,15 @@ void Fleet::Enter(const System &system, list<shared_ptr<Ship>> &ships, const Pla
 					linkVector.insert(linkVector.end(), 8, neighbor);
 			}
 		}
-	
+
```
Comment (Hurleveur) [inline source/Fleet.cpp:303]:
```suggestion

```

File: source/EscortDisplay.cpp [hunk 18]
```diff
@@ -96,7 +96,7 @@ void EscortDisplay::Draw(const Rectangle &bounds) const
 		// Draw the system name for any escort not in the current system.
 		if(!escort.system.empty())
 			font.Draw(escort.system, pos + Point(-10., 10.), elsewhereColor);
-
+
```
Comment (Hurleveur) [inline source/EscortDisplay.cpp:99]:
```suggestion

```

File: source/Command.h [hunk 19]
```diff
@@ -71,7 +71,7 @@ class Command {
 	// Modifier command, usually triggered by shift-key. Changes behaviour of
 	// other commands like NEAREST, TARGET, HAIL and BOARD.
 	static const Command SHIFT;
-
+
```
Comment (Hurleveur) [inline source/Command.h:74]:
```suggestion

```

File: source/AI.cpp [hunk 20]
```diff
@@ -3800,7 +3822,7 @@ void AI::IssueOrders(const PlayerInfo &player, const Orders &newOrders, const st
 			
 			gaveOrder = true;
 			hasMismatch |= !orders.count(ship);
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:3825]:
```suggestion

```

File: source/AI.cpp [hunk 21]
```diff
@@ -3778,7 +3800,7 @@ void AI::IssueOrders(const PlayerInfo &player, const Orders &newOrders, const st
 	// If this is a move command, make sure the fleet is bunched together
 	// enough that each ship takes up no more than about 30,000 square pixels.
 	double maxSquadOffset = sqrt(10000. * squadCount);
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:3803]:
```suggestion

```

File: source/AI.cpp [hunk 22]
```diff
@@ -1782,7 +1797,7 @@ bool AI::MoveTo(Ship &ship, Command &command, const Point &targetPosition, const
 	
 	bool shouldReverse = false;
 	dp = targetPosition - StoppingPoint(ship, targetVelocity, shouldReverse);
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:1800]:
```suggestion

```

File: source/AI.cpp [hunk 23]
```diff
@@ -1645,7 +1660,7 @@ void AI::Refuel(Ship &ship, Command &command)
 		ship.SetTargetStellar(parentTarget);
 	else if(!CanRefuel(ship, ship.GetTargetStellar()))
 		ship.SetTargetStellar(GetRefuelLocation(ship));
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:1663]:
```suggestion

```

File: source/AI.cpp [hunk 24]
```diff
@@ -398,7 +379,7 @@ void AI::UpdateKeys(PlayerInfo &player, Command &activeCommands)
 		newOrders.target = player.FlagshipPtr();
 		IssueOrders(player, newOrders, "gathering around your flagship.");
 	}
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:382]:
```suggestion

```

File: source/AI.cpp [hunk 25]
```diff
@@ -353,7 +334,7 @@ void AI::UpdateKeys(PlayerInfo &player, Command &activeCommands)
 	
 	if(!flagship || flagship->IsDestroyed())
 		return;
-
+
```
Comment (Hurleveur) [inline source/AI.cpp:337]:
```suggestion

```

File: source/AI.cpp [hunk 26]
```diff
@@ -1593,37 +1581,64 @@ void AI::MoveEscort(Ship &ship, Command &command) const
...
+			}
+			else if(ship.GetTargetStellar())
+			{
+				MoveToPlanet(ship, command);
+				if(parent.IsLanding())
+					command |= Command::LAND;
+			}
+			else if(needsFuel)
+				// Return to the system center to maximize solar collection rate.
+				MoveTo(ship, command, Point(), Point(), 40., 0.1);
```
Comment (Hurleveur) [inline source/AI.cpp:1638]: I'm not sure I agree with this.
If the player just presses land with a whormhole in the system without targeting the system on the other side, this ship should follow it into to, shouldn't it?
I haven't tried it yet, maybe it works that way.

File: source/AI.cpp [hunk 27]
```diff
@@ -1430,6 +1412,8 @@ void AI::MoveIndependent(Ship &ship, Command &command) const
 	const System *origin = ship.GetSystem();
 	if(!ship.GetTargetSystem() && !ship.GetTargetStellar() && !shouldStay)
 	{
+		// TODO: This should problably be changed, because JumpsRemaining
+		// does not return an accurate number.
```
Comment (Hurleveur) [inline source/AI.cpp:1416]: Maybe specify what needs to be changed about it? It doesn't consider whormholes?

**Review by Hurleveur — CHANGES_REQUESTED** [review]
Lets start by removing all whitespaces so we can see more clearly what's going on.
I feel like this is just an alternate approach of https://github.com/endless-sky/endless-sky/pull/6940 except it focuses more on fixing all of the bugs.
I think both PR could be complementary, with the other one focusing on performance aspect and this one bugs.

File: source/AI.cpp (hunk 26 above)
Comment (yjhn) [reply source/AI.cpp:1636]: As far as I remmeber, this block is only entered if the ship cannot land on a wormhole the parent is landing / has landed on. So the sensible thing to do in this case is to refuel if possible.

**Comment by yjhn** [comment]
Yeah, those whitespace corrections were bad. I'll fix them locally.

File: source/Ship.cpp [hunk 28]
```diff
@@ -2745,6 +2746,24 @@ double Ship::DisabledHull() const
...
+		if(p)
+			jumpFuel = JumpFuel(p->GetTargetSystem());
+	}
+	if(!jumpFuel)
+		jumpFuel = JumpFuel(targetSystem);
+	return (fuel < jumpFuel) && attributes.Get("fuel capacity") && Fuel() != 1.;
+}
+
+
+
```
Comment (warp-core) [inline source/Ship.cpp:2766]: This should go below JumpsRemaining, as it does in the header file.

**Review by warp-core — COMMENTED** [review]
@yjhn can you bring this up to date and resolve the conflicts?
It's probably best to do that before attempting any other changes.
I think most of the empty line removals in here don't need to be here, those lines help break up the code a little and make it more readable.

**Comment by yjhn** [comment]
I was planning to get back to this, but really busy right now.

**Comment by Amazinite** [comment]
We're going to be making a stable release tomorrow, and I was hoping to get this fix in since we've seen an increase in people reporting pathfinding issues around the Ember Waste as a result of this bug, so would you be able to get to this within the next day, or would you be fine with us moving the work on this over to the PR that warp-core just opened?

**Comment by yjhn** [comment]
Go on with the new PR.

**Comment by Amazinite** [comment]
Ok. Closing in favor of https://github.com/endless-sky/endless-sky/pull/7379.

**Pull request closed without merging**